File: Problem_23/main.cpp
#include <iostream>
#include <vector>

bool isAbundant(int number)
{
    int suma = 0;
    for (int i = 1; i < number / 2 + 1; ++i)
    {
        if (number % i == 0)
        {
            suma += i;
            if (suma > number)
                return true;
        }
    }
    return false;
}

bool isSum(int n, std::vector<int> abundant)
{
    std::vector<int> indexes;
    for (int i = 0; i < abundant.size(); ++i)
    {
        if (abundant[i] < n)
            indexes.push_back(i);
    }

    for (auto& i : indexes)
    {
        for (auto& j : indexes)
        {
            if (abundant[i] + abundant[j] == n)
                return true;                
        }
    }
    return false;
}

int main(int argc, char const *argv[])
{
    std::vector<int> abundantNumbers;
    for (int i = 1; i <= 28123 ; ++i)
    {
        if (isAbundant(i))
            abundantNumbers.push_back(i);
    }

    size_t total = 0;

    for (int i = 1; i < 28123; ++i)
    {
        if (!isSum(i, abundantNumbers))
            total += i;
    }
    std::cout << total << std::endl;
    return 0;
}

File: Problem_23/main.py
from tqdm import tqdm
import numpy 
from numba import njit

@njit
def isAbundant(n):
    suma = 0
    for i in range(1, n // 2 + 1):
        if (n % i == 0):
            suma += i
            if suma > n:
                return True
    return False

@njit
def isSum(num, abundant):

    indexes, = numpy.where(abundant < num)
    for i in indexes:
        for j in indexes:
            if (abundant[i] + abundant[j] == num):
                return True

    return False

def main():

    abundant_numbers = list()
    for i in tqdm(range(28123)):
        if isAbundant(i):
            abundant_numbers.append(i)

    abundant_numbers = numpy.array(abundant_numbers)

    total = 0
    for i in tqdm(range(28123)):
        if not isSum(i, abundant_numbers):
            total += i

    print (total)

if __name__ == '__main__':
    main()